### ISSUE TITLE
What to do with blogs that require tumblr login?

### ISSUE THREAD

**Issue by escaroda** [johanneszab/TumblThree#4]
I've seen some blogs that works only after I login on tumblr website.
Is there any idea how we can implement authorization?

**Comment by Ilapides** [comment]
This is a big issue for me. Is there some way to route TumblThree through an OAuth token?

**Comment by johanneszab** [comment]
I've written [a bit here](https://github.com/johanneszab/TumblThree/wiki/New-Feature-Requests-and-Possible-Enhancements) about it. The newer [Tumblr API v2](https://www.tumblr.com/docs/en/api/v2) works with OAuth. The Code section in question would be [in here](https://github.com/johanneszab/TumblThree/blob/master/src/TumblThree/TumblThree.Applications/Controllers/ManagerController.cs).

I'll probably have some time over the next weeks too, so I might look into it as well. But I've never used OAuth, so I'm new to it as well. Maybe there is already a official or similar implementation to look at around.

**Comment by johanneszab** [comment]
Check out [this wiki page](../wiki/Download-of-private-blogs----Tumblr-api-v2). Its certainly far from the optimal solution, but everything we have so far.

**Comment by johanneszab** [comment]
Well, that didn't turn out to be a good idea.

It's probably easier than I thought to do it the manual way and parse the whole website using a library like [Html Agility Pack](https://htmlagilitypack.codeplex.com/) or [AngleSharp](https://anglesharp.github.io/). We could use the internal .NET browser to open the login page and use the [cookie container](http://stackoverflow.com/questions/15049877/getting-webbrowser-cookies-to-log-in) to grab all the cookies and than use them within the WebRequests and parse the results with Html Agility Pack/AngleSharp.

It's probably worth a try and I'd try it out if no one else is willing to once I've some more spare time again.

**Comment by kingbode** [comment]
I already worked on a code for a WebSite that require login to download pages from it correctly, and it worked fine with me using cookie container and WebClient to download the pages and parse them and get images URLs and download them. as below, the issue is that looks like custom to this website , any be there are other ways which is easier and should be formalized on universal process to fit for all websites, or you can use the option login manually and open the WebBrowser control and do all your work from within it ,after you login to your site , NewDownloader application is using the same way, but IDM is more smarter and I think it has the super process to achieve this 

`
CookieAwareWebClient client = new CookieAwareWebClient();
 
public Boolean LogintoThePage(CookieAwareWebClient client) 
        {

           client.Encoding = Encoding.UTF8;

            var loginData = new NameValueCollection();
            loginData.Add("login", "YourUserName");
            loginData.Add("password", "YourPassWord");
            loginData.Add("submit", "Войти");
            client.UploadValues("http://gde-fon.com/user/login", loginData);


            return IloggedSuccessfully = true;

        }
`

**Comment by johanneszab** [comment]
So, we could also use a WebDriver, which essentially is a chrome/firefox without the UI. That would also simplify the private blog access. It probably uses lots of memory and fattens the application a lot, but it sure is the most simple implementation for now.

**Comment by johanneszab** [comment]
I've actually played around with these several times. You need ajax/XMLHttpRequest in a POST request from what I've seen in the Chrome developer tools for the pagination (i.e. get the next images). But I couldn't figure out how to properly do this within C#'s HttpWebRequest and what Headers/cookies need to be send with, how to wait for the answer, etc. Does this even work without any javascript?

**Comment by johanneszab** [comment]
Okay, I've figured it out. It will probably work with the upcoming safe mode too, and we might even drop the whole api. Would be beautiful ..

**Comment by johanneszab** [comment]
This release can download private blogs that require a login. It also sends a flag in every request to get nsfs content. For this to function it uses a different server from the Tumblr.com network and a cookie for authentication.

Until this part of [this branch](https://github.com/johanneszab/TumblThree/tree/svc) has been included into the main branch it will stay here as a separate download. It would be nice to have a proper detection that switched between the different detection method depending on its requirement. Say you don't want to download a private/nsfw blog, than you don't have to login use the cookie for every request. If the request fails, try again with a cookie..

* Should allow to bypass the [safe mode](https://tumblr.zendesk.com/hc/en-us/articles/231885248-Sensitive-content) and download nsfw/explict content after July 5.
* Allows to download private blogs (#4).
* Can grab more [meta information](https://www.jzab.de/files/svc_metadata.txt). Someone has to parse the proper parts. I'll probably not do this.
* You must login in order to do anything unlike previous releases. Even for adding a blog. Go to _Settings->Authenticate_ and fill in your username / password in the Internet Explorer window. It will automatically close after the login process and store the cookie. Alternatively, you could also open the Internet Explorer and login there.
* You can increase/remove the rate limit for crawling the website since it doesn't use the Tumblr api. But I personally wouldn't without a solid reason.
* The maximum posts per pages are now 100 posts instead of 50 as it used to be with the Tumblr api. So, I would increase the post per page count as it reduces the amount of connections necessary to crawl the blog.
* _Pre-release for a reason_: It's not well tested at all. I personally would use the [latest stable release](https://github.com/johanneszab/TumblThree/tree/v1.0.6.8), unless you want to download a private blog, or __if__ the stable release of TumblThree will not work anymore after July 5.

[TumblThree-v1.0.7.1.zip](https://github.com/johanneszab/TumblThree/files/1126792/TumblThree-v1.0.7.1.zip)
